### ISSUE TITLE
Use environment variable to restrict g_debug output

### ISSUE THREAD

**Issue by spk121** [spk121/guile-gi#58]
The `g_debug` log messages, while a bit haphazard, have been useful. However, they produce a lot of output.  They could be categorized and only produce output based on the setting of an environment variable.

Those categories could include
- function and type creation
- memory alloc/dealloc
- function calls

**Comment by LordYuuma** [comment]
Am I correct in assuming, that this would effectively map to per-file log domains? The only other solution that would come to mind would be using structured array/variant logging.

**Comment by spk121** [comment]
I was thinking like allowing the supression of some categories of information. I was imagining something like how the `LD_DEBUG` environment variable works with the libc `ld.so`  dynamic linker.

By default, if debug output is being printed, all of it will be printed.  But if, say `GI_DEBUG=memory`, only memory messages will be output.  Effectively
```
void g_mem_debug(...) { if (debug_out & MEM_FLAG)  g_debug(...); }
```

**Comment by LordYuuma** [comment]
I think implementing `g_mem_debug` (and `g_xxx_debug` in general) on top of `g_log_<something>` and letting the log handler deal with the environment variable stuff makes for a better interface. That way, (given enough documentation on the matter), a user may decide to do their own filtering, to perhaps redirect different streams to different files or what have you, in a log handler they themselves write. Of course while still supporting the `GI_DEBUG` variable.

**Comment by spk121** [comment]
Fair enough.

I think you're suggesting that `g_mem_debug` be aliased to `g_log_structured` with a an appropriate category tag, which would be easy to do.

**Comment by LordYuuma** [comment]
"Implemented in terms of" rather than "aliased", but essentially yes. I think it would be wise to use a field that feels similar to fields already used in some core libraries, such as `GIG_LOG_DOMAIN` analogous to `G_LOG_DOMAIN` (under the assumption that those fields serve a similar purpose, of course).

**Comment by LordYuuma** [comment]
I've done some preliminary work in 2371f7e22b8dd53461da0a270e47537e77dc6f3f. Now you only need to add your own functions and use them.

Changes w.r.t. our discussion:
- I used the environment variable GIG_DEBUG instead of GI_DEBUG, so as to not cause confusion with GI itself.
- I used the format `gig_debug_xxx` instead of `g_xxx_debug`, because that was already used in one of our macros.

**Comment by spk121** [comment]
Sweet

**Comment by LordYuuma** [comment]
Did some additional work in cbe94a. WDYT?

**Comment by spk121** [comment]
Looks good.  I put together some rough docs, so maybe read through them to see if they make sense.

I think it is probably time to ship a new rev as well.

**Comment by LordYuuma** [comment]
I'd call it 0.2.1 then, since it's more of a bugfix release and doesn't change much of the interface (I hope).

The journald configure flag touches on something, that I was already unsure about – whether we should even register our own writer. According to GLib, it should not be libraries, which change this setting, but applications. Before 1.0, we should merge our domains into G_LOG_DOMAIN and provide functionality to add custom loggers. (But for 0.2.1 and even 0.3 the current handling is fine imo.)

As far as the doc is concerned, you mistakenly reference the output port instead of the error port once. (You later correctly say error port.) I'm also not sure, whether the current error port affects the logging thread, or whether you need to set the default error port, but if you've tested it, that's fine.

**Comment by spk121** [comment]
OK.  I think a better way to go is to make a scheme function that takes a SCM port argument to install your logger, and another scheme function to log to systemd journal, for systemd distros.  Of course, there will be a few log messages that will occur before one can make the logger-install function call, but, I think that's OK.

That does mean that someone who enables G_MESSAGES_DEBUG without installing a either logger will be getting the full spew, but, we've provided and documented a workaround, so that's probably OK.

I made patch to do that, but, it revealed an odd unref bug I'm trying to track down.


My calendar is busy, but, I'll try to ship 0.2.1 this weekend after taking another look at the docs.

**Comment by LordYuuma** [comment]
I too have already attempted to write custom log hooks in Scheme, but it did not turn out well. However, if we make `(gi logging)` self-contained, a user can import it before any other `(gi)` module and thus setup logging before `(gi)` itself is loaded. Either way, the logging API will need some work.

Docs and NEWS should be the only missing items for this release. My last commit hopefully fixed the blocked handler, even if I myself don't fully understand it.

**Comment by spk121** [comment]
Closed by aec48208607f27d520f240f7a7104a667af15ba3
